typed-racket-lib/typed-racket/private/type-contract.rkt

```diff
@@ -412,17 +415,25 @@
         ;; Key with (cons name 'app) instead of just name because the
         ;; application of the Name is not necessarily the same as the
         ;; Name type alone
-        (cond [(hash-ref recursive-values (cons name 'app) #f)]
-              [else
-               (define name* (generate-temporary name))
-               (recursive-sc (list name*)
-                             (list
-                              (t->sc (resolve-once type)
-                                     #:recursive-values
-                                     (hash-set recursive-values
-                                               (cons name 'app)
-                                               (recursive-sc-use name*))))
-                             (recursive-sc-use name*))])]
+        (define constr (lookup-type-constructor name))
+        (cond
+          ;; when constr is a built-in or non-recursive user-defined type
+          ;; constructor, don't generate a recursive static contract
+          ;; for the resulting type.
+          [(not (and (user-defined-type-constr? constr)
```

Comment:
I would move this expression and the `lookup-type-constructor` call into a function like `simple-type-constructor?`.